keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1003 @@
+abstraction is expected to be the result of applying Kubernetes' `Service` semantics and business
+logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business
+  logic changes
+- provide a client library with minimal dependencies
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- create reusable library elements for some proxy logic, such as conntrack cleaning, which might be called from different proxy implementations
+
+### Non-Goals
+
+- provide equivalent implementations of all third-party proxiers
+- force new users to run kpng in two separate daemons - this has performance benefits and is optional
+- force all proxiers to use the fullstate model: this is ideal for new implementations (easier to understand, no magic caches) but its not necessary.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible as usual via
+TCP (`127.0.0.1:12345`) and/or via a socket (`unix:///path/to/proxy.sock`).
+
+- it will connect to the API server and watch resources, like the current proxy;
+- then, it will process them, applying Kubernetes specific business logic like topology computation
+  relative to the local host;
+- finally, provide the result of this computation to client via a gRPC watchable API.
+
+This decoupling allows kube-proxy and implementation to evolve in their own timeframes. For
+instance, introducing optimizations like EndpointSlice or new business semantics like Topology
+does not trigger a rebuild/release of any proxy implementation.
+
+In this implementation, we either: 
+- send the *full state*  of the kubernetes networking state-space to a client, every time anything needs to change.  Since this is done over GRPC or in memory, the bandwidth costs are low (anecdotally this has been measured, and it works for 1000s of serviecs and pods - we can attach specific results to this KEP as needed).  An example of this can be seen in the ebpf and nft proxies in the KPNG project (https://github.com/kubernetes-sigs/kpng/blob/master/backends/nft/nft.go).
```

Comment:
(typo "serviecs")

You need to include your data here.

Reply:
note full state maintains the full state on the client side, so it's only done in memory, not over gRPC (gRPC always carry the diffs only).

Reply:
@mcluseau , do you have a link to the experiments  you ran ? or maybe we can build a cluster and do a one time experiment ?